drivers/arm/ws2812.c

```diff
@@ -0,0 +1,111 @@
+#include "quantum.h"
+#include "ws2812.h"
+#include "ch.h"
+#include "hal.h"
+
+#if !defined(LED_ARRAY) && defined(RGB_MATRIX_ENABLE)
+// LED color buffer
+LED_TYPE led[DRIVER_LED_TOTAL];
+#    define LED_ARRAY led
+#endif
+
+#ifdef RGB_MATRIX_ENABLE
+// Set an led in the buffer to a color
+void inline ws2812_setled(int i, uint8_t r, uint8_t g, uint8_t b) {
+    led[i].r = r;
+    led[i].g = g;
+    led[i].b = b;
+}
+
+void ws2812_setled_all(uint8_t r, uint8_t g, uint8_t b) {
+    for (int i = 0; i < sizeof(led) / sizeof(led[0]); i++) {
+        led[i].r = r;
+        led[i].g = g;
+        led[i].b = b;
+    }
+}
+#endif
+
+/* adapted from
+https://github.com/bigjosh/SimpleNeoPixelDemo/blob/master/SimpleNeopixelDemo/SimpleNeopixelDemo.ino
+*/
+
+#define NS_PER_SEC (1000000000L)  // Note that this has to be SIGNED since we want to be able to check for negative values of derivatives
+
+#define CYCLES_PER_INSTRUCTION 0.35
+
+#define CYCLES_PER_SEC (STM32_SYSCLK * CYCLES_PER_INSTRUCTION)
+
+#define NS_PER_CYCLE (NS_PER_SEC / CYCLES_PER_SEC)
+
+#define NS_TO_CYCLES(n) ((n) / NS_PER_CYCLE)
+
+#define wait_ns(x)                              \
+    for (int i = 0; i < NS_TO_CYCLES(x); i++) { \
+        __NOP();                                \
+    }
```

Comment:
Maybe enclose this block in a `do { ... } while(0)`, so it can be used in single-statement conditionals and the like

Reply:
also, would it be worth moving ns-accurate delay code to a more common arm thing?

Reply:
actually, what about @skullydazed PR #7124 instead?

Reply:
That PR only handles micro not nano second delays. I think I would prefer to have another consumer before moving to core, as its really something to live in "wait.h" and have implementations for all platforms.